apphub/instance_segmentation/solov2/solov2_tf.py

```diff
@@ -0,0 +1,574 @@
+                    "category_id": self.mapping[label],
+                    "score": score,
+                    "segmentation": mask_util.encode(np.array(seg[..., np.newaxis], order='F'))[0]
+                }
+                self.results.append(result)
+        return data
+
+    def on_epoch_end(self, data):
+        mAP = 0.0
+        if self.results:
+            with Suppressor():
+                coco_results = self.coco_gt.loadRes(self.results)
+                cocoEval = COCOeval(self.coco_gt, coco_results, 'segm')
+                cocoEval.evaluate()
+                cocoEval.accumulate()
+                cocoEval.summarize()
+                mAP = cocoEval.stats[0]
+        data.write_with_log(self.outputs[0], mAP)
+
+
+def get_estimator(data_dir=None,
+                  epochs=12,
+                  batch_size_per_gpu=8,
+                  im_size=1024,
+                  model_dir=tempfile.mkdtemp(),
+                  train_steps_per_epoch=None,
+                  eval_steps_per_epoch=None):
+    assert im_size % 32 == 0, "im_size must be a multiple of 32"
+    num_device = get_num_devices()
+    train_ds, val_ds = mscoco.load_data(root_dir=data_dir, load_masks=True)
+    batch_size = num_device * batch_size_per_gpu
+    pipeline = fe.Pipeline(
+        train_data=train_ds,
+        eval_data=val_ds,
+        test_data=val_ds,
+        batch_size=num_device * batch_size_per_gpu,
+        ops=[
+            ReadImage(inputs="image", outputs="image"),
+            MergeMask(inputs="mask", outputs="mask"),
+            GetImageSize(inputs="image", outputs="imsize", mode="test"),
+            LongestMaxSize(max_size=im_size, image_in="image", mask_in="mask", bbox_in="bbox", bbox_params="coco"),
+            PadIfNeeded(min_height=im_size,
+                        min_width=im_size,
+                        image_in="image",
+                        mask_in="mask",
+                        bbox_in="bbox",
+                        bbox_params="coco",
+                        border_mode=cv2.BORDER_CONSTANT,
+                        value=0),
+            Sometimes(HorizontalFlip(image_in="image", mask_in="mask", bbox_in="bbox", bbox_params="coco",
+                                     mode="train")),
+            Resize(height=im_size // 4, width=im_size // 4, image_in='mask'),  # downscale mask for memory efficiency
+            Gt2Target(inputs=("mask", "bbox"), outputs=("gt_match", "mask", "classes")),
+            Delete(keys="bbox"),
+            Delete(keys="image_id", mode="!test")
+        ],
+        pad_value=0,
+        num_process=8 * num_device)
+    init_lr = 1e-2 / 16 * batch_size
+    model = fe.build(model_fn=lambda: solov2(input_shape=(im_size, im_size, 3)),
+                     optimizer_fn=lambda: tf.optimizers.SGD(learning_rate=init_lr, momentum=0.9))
+    network = fe.Network(ops=[
+        Normalize(inputs="image", outputs="image", mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)),
```

Comment:
why not the Numpy normalize?

Reply:
it's faster to do it on gpu, because the cpu- gpu transfer type is uint8 instead of float32